// config/test/config/default.js
"use strict";

// Default configuration values.  These values apply to all environments unless they are
// overridden by environment-specific configuration files.

module.exports = {

    // MongoDB-specific configuration.
    mongodb : {

        serverConfig : {
            auto_reconnect:true,
            poolSize:1,
            slave_ok:true,
            safe:true
        },
        dbConfig : {
            forceServerObjectId: false,
            native_parser: false
        },
    },

    // MySQL-specific configuration.
    mysql : {

        dbName : 'my_game_db',
        dbPass : '',
        dbHost : 'localhost',
        dbUser : 'root',
        maxPoolSize : 10,   // maximum number of db connections to create at any given time
        idleTimeoutMillis : 30000,  //milliseconds a resource can go unused before it should be destroyed (default 30000)
    },

    memcached : {
        serverLocations : [ 'localhost:11211' ],
        options : { "poolSize" : 5 }
    },

    analytics : {

        ngPipes : {
            host : 'ngpipes-sandbox.mobage.com',
            path : '/pipes/r.2/bulk_record_stats'
        }
    },

    region : "US",
    mobage : {},
    phone : {},
    gameServer : {},
    SocialURL : {},

    US : {

        testUser : {
            userId : 7202,
            gamertag : 'jcw01',
            password : 'jcw01',
            udid :  '38F783D8-F7A7-5C4C-99E7-C0F0F74F8442'
        },

        mobage : {
            enabled : false,
            api_version : 1,
            host : 'app-sandbox.mobage.com',
            bankHostName : 'bank-sandbox.mobage.com',
            appKey : "ngCoreServer",
            key :  '27DXBcnHjI3PJ5erek9mag',
            secret : 'noGbiaXzhFjmBJ6rrjg7zFSGKrnj9WlXYhwkhNVvoZQ',
            consumerKey : 'f2iKusPFUk23MW22aSFg',  // actually is clientKey
            consumerSecret : 'IGYX2F4hNU9Qhq2OEcWTm9j3QtRcffwVLaxwXYUQ0',  // actually is clientSecret
            clientKey : 'f2iKusPFUk23MW22aSFg',
            clientSecret : 'IGYX2F4hNU9Qhq2OEcWTm9j3QtRcffwVLaxwXYUQ0',
        },

        phone : {

            api_version : 1,
            host : 'app-sandbox.mobage.com',
            appKey : 'ngCoreServer',
            consumerKey : 'f2iKusPFUk23MW22aSFg',
            consumerSecret : 'IGYX2F4hNU9Qhq2OEcWTm9j3QtRcffwVLaxwXYUQ0',
        },

        gameServer : {
            api_version : 1,
            host : 'localhost',
            port : 8090,
            // bindToAddress: '10.16.101.118', // Leave commented out for INADDR_ANY
            appKey : 'ngCoreServer',
            consumerKey : 'f2iKusPFUk23MW22aSFg',  // actually is clientKey
            consumerSecret : 'IGYX2F4hNU9Qhq2OEcWTm9j3QtRcffwVLaxwXYUQ0',  // actually is clientSecret
        },

        SocialURL : "opensocial"
    },

    JP : {
        testUser : {
            userId : 7202,
            gamertag : 'jcw01',
            password : 'jcw01',
            udid :  '38F783D8-F7A7-5C4C-99E7-C0F0F74F8442'
        },

        mobage : {
            enabled : false,
            auth_api_version : "v2.01",
            bank_api_version : "v2.01",
            social_api_version : "v2",
            host : 'sb.sp.mbga-platform.jp',
            bankHostName : 'ssl.sb.sp.mbga-platform.jp',
            appKey : "",
            key :  '',
            secret : '',
            consumerKey : '',  // actually is clientKey
            consumerSecret : '',  // actually is clientSecret
            clientKey : '',
            clientSecret : '',
        },

        phone : {

            api_version : 1,
            host : '',
            appKey : '',
            consumerKey : '',
            consumerSecret : '',
        },

        gameServer : {
            api_version : 1,
            host : 'localhost',
            port : 8090,
            // bindToAddress: '10.16.101.118', // Leave commented out for INADDR_ANY
            appKey : '',
            consumerKey : '',  // actually is clientKey
            consumerSecret : '',  // actually is clientSecret
        },

        SocialURL : "social"
    },
    
    /**
     * logging.pattern special variables. See log4js docs for more details.
     *
     * NOTE: logging.pattern currently only affects our file appender.
     *
     *  - %r time in toLocaleTimeString format
     *  - %p log level
     *  - %c log category
     *  - %m log data
     *  - %d date in various formats
     *  - %% %
     *  - %n newline
     */
    logging : {
        //logfile : '/tmp/ngserver.log',  // optional logfile override, default is {appHome}/logs/{appEnv}.log
        logLevel : 'INFO',
        pattern : "[%d{ISO8601}] [%p] %m{1}",
        logRequests : false,
        logResponses: false,
        logBodyMax: 2048,
    },

    exitOnUncaughtException : false,

    metrics : {

        collectHttpMetrics : true,
        collectMongoDbMetrics : true,
        collectMemcachedMetrics : true,
        collectMysqlMetrics : true
    },

    long_stack_traces: false,

    foo: "bar",
};
